doc/architecture-decisions/0018-route-crd-and-contour-controller.md

```diff
@@ -0,0 +1,48 @@
+# 18. Add Contour as an Alternative Ingress Solution Provider
+
+Date: 2020-10-09
+
+## Status
+
+Proposal
+
+## Context
+In our efforts to allow more optionality in ingress solutions for CF-for-K8s, we
+want to allow the use of Contour as a potential alternative to Istio.
+
+### Proposed Design
+
+We propose:
+1. Adding a configuration option `ingress_solution_provider` to CF-for-K8s, with
+   potential values `istio` or `contour`.
+2. Creating another kubebuilder controller, similar to routecontroller, to
```

Comment:
Why a separate controller instead of swapping out the `VirtualService` resource builder for an `HTTPProxy` builder?

Not doubting you have good reasons. It's just unclear from the content of this ADR.

Reply:
I am also curios why a separate controller

Reply:
I think it'll be simpler to maintain two separate ones, and I also worry about the behavior of a Controller that tries to assert control over CRDs that don't exist in a cluster (presumably we wouldn't push the definition for VirtualServices on a contour cluster). I haven't investigated anything though.

Reply:
Hey @tcdowney, good question! 

We did consider having a single `Routecontroller` reconcile either a `VirtualService` or a `HTTPProxy` resource depending on the ingress provider an operator chooses. However, a Kubernetes controller can/should only handle a single CRD ([source](https://github.com/kubernetes-sigs/kubebuilder/issues/581)) so we've decided to have two controllers, one reconciling HTTPProxy CRs and one reconciling VirtualService CRs.

We will add a bit more context explaining why we chose to have 2 controllers instead of one though :)

Reply:
@kauana is it possible to have two entrypoints for every CRD but share the rest of the code?

Reply:
Isn't the RouteController reconciling Route CRs and not either VirtualService or HTTPProxy?

Reply:
> However, a Kubernetes controller can/should only handle a single CRD (source) 

I see that link mentions:
> Best practice is to have one controller reconciling objects of one type. 

My understanding was that the routecontroller only watches/reconciles a single type, the `Route` CRD. So this wouldn't violate that.

Unless maybe it is configured to watch child resources as well for changes? I don't think it does, though:
https://github.com/cloudfoundry/cf-k8s-networking/blob/c1a8c866375fd964f08cc5613fced88e572eefdb/routecontroller/controllers/networking/route_controller.go#L254

Reply:
Yes, we do reconcile Route CRs and create/update/delete Virtual Services and Services accordingly. My point was more towards having a single RouteController reconciling both Virtual Service and HTTPProxy resources, two completely different CRDs, when in reality an operator must choose a single solution provider (contour or istio). 

In my opinion, the reconciliation loop should be separated into two controllers because it makes our controller(s) logic much simpler and easier to understand. For example, if an operator wants to use Istio, there is no need to have any added overhead/complexity added to the controller to try reconciling HTTPProxies at all, and by having the 2 controllers and a single controller deployed at a time, we can accomplish the same goal without any added complexity (+ debugging should be easier too since the controller logic is split up :upside_down_face: )

Reply:
I wrote this ADR with Kauana, but after looking more at the routecontroller code, and some of the previous comments, I think I'd prefer it to be in one controller also.

I didn't really understand routecontroller's relationship with route CRDs vs virtualservices. As pointed out in a couple comments above, routecontroller only reconciles route CRs. In this instance, we only wanna change what it produces as a result of that reconciliation (virtualservices or httpproxies).

I now think that it makes more sense to just add another resource builder and just select the builder via routecontroller config on startup. It feels simpler.

@XanderStrike @mike1808 curious if you have any other thoughts.

Reply:
My only two issues are:

1. Can RouteController be deployed on environments where VirtualServices don't exist? I know it doesn't watch them, but I don't know the full machinations of what KubeBuilder does when we tell it that it should own a type of resource, I worry it validates that the resources it knows about are available on the cluster it's deployed to. This should be quick to test.
2. If we use one RC then basically all of the important business logic in it will be wrapped in if statements. My gut feeling is this is gonna make things more complicated and harder to read than if we had these two totally separate code paths living in two totally separate components. I think we could mitigate these concerns by branching as early as possible and keeping the code well compartmentalized, maybe by having different packages for VS and HTTPProxy. My main goal would be to avoid all our files being littered with "if istio else contour."

Reply:
Personally, I still believe it should be 2 different controllers (2 different deployments). It is true we only reconcile the route CR, but our controller has to translate that into completely different CRDs and since we are using kubebuilder, if the controller "speaks" both VS and HTTPProxy, you will need to install both CRDs into the cluster.  Then, AFAIK if an operator is using Contour and we have a single controller, then the operator will be able to try getting "Virtual Service" resources because the CRD is installed by kubebuilder into the cluster and an operator may be confused because they are not using Istio and yet they have the VirtualService CRD.

I also agree with Alex that by having a single controller will make things more complicated for us and the code may look a bit ugly. Also, right now we only support Contour and Istio, but what if we end up supporting more Ingress solutions (e.g AVI, OSM, etc)? Having a controller reconcile all these resources will keep adding more and more complexity to the controller, which could be avoided by implementing these solutions in parallel.

Lastly, I think implementing the 2 versions of the routecontroller should be straightforward since the istio controller is already written and we could simply clone the routecontroller and use http proxy resource builder instead (Mike has done this as part of an exploration I believe).